drivers: led: lp5569: implement write_channels api

Emil-Juhl opened the pull request:
[drivers: led: lp5569: enable auto-increment](https://github.com/zephyrproject-rtos/zephyr/commit/574c4a5491b4e6e93c7c93c0242b50184de102fd)

The lp5569 controller supports auto-increment for its registers. This is by
default enabled in the chip, but the driver actively disabled it.

Since auto-increment is useful for writing multiple consecutive registers,
enable the feature. The driver, however, doesn't perform such consecutive
writes and thus the existing behavior is not altered.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

[drivers: led: lp5569: implement write_channels api](https://github.com/zephyrproject-rtos/zephyr/commit/cff92077db2e05b0aa30ff60c8ccef68beeaa525)

The lp5569 has multiple pwm outputs, and thus implementing the
write_channels api to set multiple values in a single call makes sense.

Implement the write_channels function with a basic range check on the
channel range.
Since the lp5569 supports auto-increment, all of the channels can be
written in one i2c transfer, starting from the pwm register of the
start_channel.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

Commit:
drivers: led: lp5569: enable auto-increment

The lp5569 controller supports auto-increment for its registers. This is by
default enabled in the chip, but the driver actively disabled it.

Since auto-increment is useful for writing multiple consecutive registers,
enable the feature. The driver, however, doesn't perform such consecutive
writes and thus the existing behavior is not altered.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

Commit:
drivers: led: lp5569: implement write_channels api

The lp5569 has multiple pwm outputs, and thus implementing the
write_channels api to set multiple values in a single call makes sense.

Implement the write_channels function with a basic range check on the
channel range.
Since the lp5569 supports auto-increment, all of the channels can be
written in one i2c transfer, starting from the pwm register of the
start_channel.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

jonas-rem reviewed (commented):
Could you also adapt the test for the driver so that this feature is used there? Some of the other tests for the led drivers already test the write_channels function.

Emil-Juhl commented:
> Could you also adapt the test for the driver so that this feature is used there?
>
I haven't been able to find any existing test for the ti,lp5569 except for `tests/drivers/build_all/led/src/main.c` which does essentially nothing.
Could I ask you to point me to the testcode that you refer to? Thanks!

> Some of the other tests for the led drivers already test the write_channels function.
>
Do you have a reference on that - again I wasn't really able to find anything. Though for this one, the led-api test, `tests/drivers/led/led_api/src/test_led_api.c`, seems relevant. It doesn't excercise the `write_channels` api though, but it could be extended to do so I guess. Is that, what you're going for?

jonas-rem commented:
> > Could you also adapt the test for the driver so that this feature is used there?
> 
> I haven't been able to find any existing test for the ti,lp5569 except for `tests/drivers/build_all/led/src/main.c` which does essentially nothing. Could I ask you to point me to the testcode that you refer to? Thanks!
> 
> > Some of the other tests for the led drivers already test the write_channels function.
> 
> Do you have a reference on that - again I wasn't really able to find anything. Though for this one, the led-api test, `tests/drivers/led/led_api/src/test_led_api.c`, seems relevant. It doesn't excercise the `write_channels` api though, but it could be extended to do so I guess. Is that, what you're going for?

Oh, it's not a test but a sample. You can find the sample for the lp5569 here:

`samples/drivers/led/lp5569/`

And for the other led drivers in the same head folder:
```
ls samples/drivers/led/
apa102c_bitbang  is31fl3194   is31fl3733  lp3943  lp5562  pca9633  sx1509b_intensity
index.rst        is31fl3216a  led_strip   lp50xx  lp5569  pwm      xec
```

Commit:
samples: led: lp5569: demo write_channels

Add a simple demonstration of the led_write_channels api on the lp5569
driver sample.
The demonstration simply turns on all of the channels with a single call to
led_write_channels. Then the same is done for turning off the channels.
Thus, it doesn't add much visually, but it shows the usage of the api.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

Emil-Juhl commented:
> Oh, it's not a test but a sample. You can find the sample for the lp5569 here:
> 
> `samples/drivers/led/lp5569/`
> 
Thank you!
I've added usage of the led_write_channels api in https://github.com/zephyrproject-rtos/zephyr/pull/81497/commits/8365651526839e0cb60176bd9ac16cdf3524531f 
If this is not what you had in mind, let me know and we'll work it out :)

jonas-rem commented:
> > Oh, it's not a test but a sample. You can find the sample for the lp5569 here:
> > `samples/drivers/led/lp5569/`
> 
> Thank you! I've added usage of the led_write_channels api in [8365651](https://github.com/zephyrproject-rtos/zephyr/commit/8365651526839e0cb60176bd9ac16cdf3524531f) If this is not what you had in mind, let me know and we'll work it out :)

Thanks! I think this is sufficient to demonstrate. I have a board with two lp5569 here and will test later today.

simonguinot commented on drivers/led/lp5569.c:
```diff
@@ -77,6 +77,23 @@ static inline int lp5569_led_off(const struct device *dev, uint32_t led)
+{
+	const struct lp5569_config *config = dev->config;
+	uint32_t i2c_len = num_channels + 1;
+	uint8_t i2c_msg[i2c_len];
```
Please avoid such uncontrolled / dynamic allocations on the stack. It makes me really nervous :) At the very least you need to limit the buffer size...

simonguinot commented on samples/drivers/led/lp5569/src/main.c:
```diff
@@ -59,6 +60,26 @@ int main(void)
+			return ret;
+		}
+
+		k_sleep(DELAY_TIME_ON);
```
You need to update the README.rst file with the new behavior.

And finally you also need to document this new feature in the release-notes.

Emil-Juhl replied:
Of course, sorry for missing this. I'll try to document it appropriately. Thanks!

simonguinot reviewed (changes requested):
Hello @Emil-Juhl,

Thanks for implementing the write_channels() API for the lp5569 driver. It is clear and straightforward.

See below for the changes I am requesting.

Emil-Juhl commented:
@simonguinot Thanks for the comments, I've resolved the conversations - but if you don't think I've resolved the issues appropriately, just re-open them and let me know.

simonguinot reviewed (approved):
Thanks @Emil-Juhl !
